File: src/main/java/org/jetbrains/research/groups/ml_methods/code_completion_benchmark/core/counter/Counter.java
package org.jetbrains.research.groups.ml_methods.code_completion_benchmark.core.counter;

import java.io.Externalizable;
import java.util.List;

public interface Counter extends Externalizable {

    int getCount();

    long[] getCounts(List<Integer> indices);

    int getCountOfCount(int n, int count);

    int getSuccessorCount();

    int getSuccessorCount(List<Integer> indices);

    List<Integer> getTopSuccessors(List<Integer> indices, int limit);

    int[] getDistinctCounts(int range, List<Integer> indices);

    void count(List<Integer> indices);

    void unCount(List<Integer> indices);

    default void countBatch(List<List<Integer>> indices) {
        indices.forEach(this::count);
    }

    default void unCountBatch(List<List<Integer>> indices) {
        indices.forEach(this::unCount);
    }
}

File: src/main/kotlin/org/jetbrains/research/groups/ml_methods/code_completion_benchmark/core/counting/Counter.kt
package org.jetbrains.research.groups.ml_methods.code_completion_benchmark.core.counting

import java.io.Externalizable
import java.util.function.Consumer

interface Counter : Externalizable {

    val count: Int

    val successorCount: Int

    fun getCounts(indices: List<Int>): LongArray

    /**
     * Returns the number of sequences of length n seen `count' times
     */
    fun getCountOfCount(n: Int, count: Int): Int

    fun getSuccessorCount(indices: List<Int>): Int
    fun getTopSuccessors(indices: List<Int>, limit: Int): List<Int>

    fun getDistinctCounts(range: Int, indices: List<Int>): IntArray

    fun count(indices: List<Int>)
    fun unCount(indices: List<Int>)

    fun countBatch(indices: List<List<Int>>) {
        indices.forEach(Consumer<List<Int>> { idx -> this.count(idx) })
    }

    fun unCountBatch(indices: List<List<Int>>) {
        indices.forEach(Consumer<List<Int>> { idx -> this.unCount(idx) })
    }
}